service/matching/version_sets.go

```diff
@@ -87,68 +114,86 @@ func depthLimiter(g *persistence.VersioningData, maxDepth int) *workflowservice.
 //
-// Deletions are not permitted, as inserting new versions can accomplish the same goals with less complexity. However,
-// sets may be dropped when the number of sets limit is reached. They are dropped oldest first - the current default set
-// is never dropped, instead dropping the next oldest set.
-func UpdateVersionSets(existingData *persistence.VersioningData, req *workflowservice.UpdateWorkerBuildIdCompatibilityRequest, maxSize int) error {
-	err := updateImpl(existingData, req)
+// Deletions are performed by a background process which verifies build IDs are no longer in use and safe to delete (not yet implemented).
+//
+// Update may fail with FailedPrecondition if it would cause exceeding the supplied limits.
+func UpdateVersionSets(clock clockpb.HybridLogicalClock, data *persistence.VersioningData, req *workflowservice.UpdateWorkerBuildIdCompatibilityRequest, maxSets, maxBuildIDs int) (clockpb.HybridLogicalClock, *persistence.VersioningData, error) {
+	clock = generateHLCTimestamp(clock)
+	data, err := updateImpl(clock, data, req)
 	if err != nil {
-		return err
+		return clock, nil, err
 	}
-	// Limit graph size if it's grown too large
-	newResp := depthLimiter(existingData, maxSize)
-	existingData.VersionSets = newResp.GetMajorVersionSets()
-	return nil
+	if err := checkLimits(data, maxSets, maxBuildIDs); err != nil {
+		return clock, nil, err
+	}
+	return clock, data, nil
 }
 
 //nolint:revive // cyclomatic complexity
-func updateImpl(existingData *persistence.VersioningData, req *workflowservice.UpdateWorkerBuildIdCompatibilityRequest) error {
+func updateImpl(timestamp clockpb.HybridLogicalClock, existingData *persistence.VersioningData, req *workflowservice.UpdateWorkerBuildIdCompatibilityRequest) (*persistence.VersioningData, error) {
 	// First find if the targeted version is already in the sets
 	targetedVersion := extractTargetedVersion(req)
 	findRes := findVersion(existingData, targetedVersion)
 	targetSetIdx, versionInSetIdx := findRes.setIdx, findRes.indexInSet
+	modifiedData := persistence.VersioningData{
+		VersionSets:            make([]*persistence.CompatibleVersionSet, len(existingData.GetVersionSets())),
+		DefaultUpdateTimestamp: existingData.GetDefaultUpdateTimestamp(),
+	}
+	copy(modifiedData.VersionSets, existingData.GetVersionSets())
 
 	if req.GetAddNewBuildIdInNewDefaultSet() != "" {
 		// If it's not already in the sets, add it as the new default set
 		if targetSetIdx != -1 {
-			return serviceerror.NewInvalidArgument(fmt.Sprintf("version %s already exists", targetedVersion))
+			return nil, serviceerror.NewInvalidArgument(fmt.Sprintf("version %s already exists", targetedVersion))
 		}
 
-		existingData.VersionSets = append(existingData.GetVersionSets(), &taskqueuepb.CompatibleVersionSet{
-			BuildIds: []string{targetedVersion},
+		modifiedData.VersionSets = append(modifiedData.VersionSets, &persistence.CompatibleVersionSet{
+			SetIds:   []string{targetedVersion},
```

Comment:
use a base64 or hex encoded hash of the version here

Reply:
Why?

Reply:
If it's just for the character limitation when creating the internal queue name, let's handle that there and keep this structure more readable.

Reply:
it is for that, and I agree it makes sense to do that closer to the use, but I would still rather do it here for a few reasons:
- if we do it here, that allows changing the hash function later. otherwise we'd have to ensure we keep the same function forever.
- for debugging purposes, I want build ids and set ids to be clearly visually distinct, since they're separate types and should not be mixed. if I see something that looks like a build id where I expect a set id that's confusing. (i.e. I believe it's _more_ readable to use a hash here.)
- if someone uses very long build ids, we do have to store them, but I'd rather store them once and store a fixed-length thing as the set id. if we use it literally we have to store it twice.

Reply:
> it is for that, and I agree it makes sense to do that closer to the use, but I would still rather do it here for a few reasons:
> 
> * if we do it here, that allows changing the hash function later. otherwise we'd have to ensure we keep the same function forever.

Fair point, it would be not impossible but highly annoying to change later with this format.
If you think that's an important capability, let's do that.

> * for debugging purposes, I want build ids and set ids to be clearly visually distinct, since they're separate types and should not be mixed. if I see something that looks like a build id where I expect a set id that's confusing. (i.e. I believe it's _more_ readable to use a hash here.)

I'd argue otherwise, that it's easier to debug which build ID the set ID correlates to and might help debug replication and task queue user data scavenger issues.

> * if someone uses very long build ids, we do have to store them, but I'd rather store them once and store a fixed-length thing as the set id. if we use it literally we have to store it twice.

Since we limit to 10 sets per queue, I'd say this is a non issue and if we really cared about this, we could optimize to omit the set ID if it's the same as the first build ID but I'd rather avoid that.


Overall I'd be okay hashing here, if you don't mind, I'll merge this and track so we don't forget.